File: 0739-Daily-Temperatures/lc0739.cc
// C++ solution.

// monostack method.
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& T) {
        // the temperature with its index
        stack<pair<int, int>> monostack;
        int num_days = T.size();
        vector<int> ans(num_days, 0);
        for (int i = 0; i < num_days; i++) {
            int curr_val = T[i];
            while (!monostack.empty() && monostack.top().first < curr_val) {
                int top_idx = monostack.top().second;
                ans[top_idx] = i - top_idx;
                monostack.pop();
            }
            monostack.push(make_pair(curr_val, i));
        }
        return ans;
    }
};

File: 0739-Daily-Temperatures/lc0739.go
// Go solution.

type TmpPoint struct {
    Val int
    Idx int
}

type TmpPointStack struct {
    Data []TmpPoint
}

func (stk *TmpPointStack) top() TmpPoint {
    return stk.Data[len(stk.Data)-1]
}

func (stk *TmpPointStack) topVal() int {
    return stk.Data[len(stk.Data)-1].Val
}

func (stk *TmpPointStack) topIdx() int {
    return stk.Data[len(stk.Data)-1].Idx
}

func (stk *TmpPointStack) push(ts TmpPoint) {
    stk.Data = append(stk.Data, ts)
}

func (stk *TmpPointStack) pop() {
    stk.Data = stk.Data[:len(stk.Data)-1]
}

func (stk *TmpPointStack) size() int {
    return len(stk.Data)
}

func dailyTemperatures(temperatures []int) []int {
    if temperatures == nil {
        return temperatures
    }

    lenIn := len(temperatures)
    ans := make([]int, lenIn)
    monoStack := &TmpPointStack{}

    for i := 0; i < lenIn; i++ {
        ithTmp := temperatures[i]

        for monoStack.size() > 0 && ithTmp > monoStack.topVal() {
            topItemIdx := monoStack.topIdx()
            ans[topItemIdx] = i - topItemIdx
            monoStack.pop()
        }
        monoStack.push(TmpPoint{ithTmp, i})
    }
    return ans
}